python/python/tests/test_async_client.py

```diff
@@ -302,6 +302,74 @@ async def test_statistics(self, glide_client: TGlideClient):
         assert "total_clients" in stats
         assert len(stats) == 2
 
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_connection_timeout(
+        self,
+        request,
+        cluster_mode: bool,
+        protocol: ProtocolVersion,
+    ):
+
+        client = await create_client(
+            request,
+            cluster_mode,
+            # addresses=multiple_replicas_cluster.nodes_addr,
+            protocol=protocol,
+            timeout=2000,
+            connection_timeout=2000,
+        )
+
+        assert isinstance(client, (GlideClient, GlideClusterClient))
+
+        assert await client.set("key", "value") == "OK"
+
+        await client.close()
+
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_connection_timeout_when_running_long_script(
+        self,
+        request,
+        cluster_mode: bool,
+        protocol: ProtocolVersion,
+    ):
+        # Create the first client to run a long-running Lua script
+        long_script_client = await create_client(
+            request,
+            cluster_mode,
+            protocol=protocol,
+            timeout=10000,  # 10 seconds timeout for the script execution
+        )
+
+        # Define a long-running Lua script (e.g., sleeps for 10 seconds)
+        long_script_code = create_long_running_lua_script(5)
+
+        # Function to run the long-running script
+        async def run_long_script():
+            await long_script_client.invoke_script(Script(long_script_code))
+
+        async def connect_to_client():
+            # Create a second client with a connection timeout of 5 seconds
+            timeout_client = await create_client(
+                request,
+                cluster_mode,
+                protocol=protocol,
+                timeout=5000,  # General timeout for operations
+                connection_timeout=7000,  # 5-second connection timeout
+            )
+
+            # Ensure the second client can connect and perform a simple operation
+            assert await timeout_client.set("key", "value") == "OK"
+
+            await timeout_client.close()
+
+        # Run the long script and attempt to connect concurrently
+        await asyncio.gather(run_long_script(), connect_to_client())
```

Comment:
this is intricate - need to ensure run_long_script runs before connect_to_client()